File: app/src/main/java/com/vladislav/workoutassistant/ui/exercises/adapters/ExerciseAdapter.java
package com.vladislav.workoutassistant.ui.exercises.adapters;

import android.util.SparseIntArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.vladislav.workoutassistant.R;
import com.vladislav.workoutassistant.utilities.DiffUtilCallback;
import com.vladislav.workoutassistant.ui.main.interfaces.OnItemClickCallback;
import com.vladislav.workoutassistant.data.db.entity.Exercise;
import com.vladislav.workoutassistant.databinding.ItemExerciseBinding;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

public class ExerciseAdapter extends RecyclerView.Adapter<ExerciseAdapter.ExerciseViewHolder>{

    private List<Exercise> mExercises;
    private SparseIntArray mSelectedExercises;
    private OnItemClickCallback mCallback;
    private int mLastClickedItemPosition;

    public ExerciseAdapter(OnItemClickCallback callback, SparseIntArray selectedExercises) {
        mCallback = callback;
        mSelectedExercises = selectedExercises;
    }

    public void updateList(List<Exercise> exercises) {
        if (mExercises == null) {
            mExercises = exercises;
            notifyItemRangeInserted(0, exercises.size());
        } else {
            DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtilCallback(mExercises, exercises));
            mExercises = exercises;
            diffResult.dispatchUpdatesTo(this);
        }
    }

    public void setLastClickedItemPosition(int position) {
        mLastClickedItemPosition = position;
    }

    public int getLastClickedItemPosition() {
        return mLastClickedItemPosition;
    }

    @Override
    @NonNull
    public ExerciseAdapter.ExerciseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemExerciseBinding binding = DataBindingUtil.inflate(inflater, R.layout.item_exercise, parent, false);
        return new ExerciseAdapter.ExerciseViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ExerciseAdapter.ExerciseViewHolder holder, int position) {
        holder.bind(mExercises.get(position));
    }

    @Override
    public int getItemCount() {
        return mExercises == null ? 0 : mExercises.size();
    }

    class ExerciseViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private ItemExerciseBinding mBinding;
        private Exercise mExercise;

        ExerciseViewHolder(ItemExerciseBinding binding) {
            super(binding.getRoot());

            mBinding = binding;
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            mLastClickedItemPosition = getAdapterPosition();
            mCallback.onClick(mExercise.getId(), mExercise.getName());
        }

        void bind(Exercise exercise) {
            mExercise = exercise;
            mBinding.setName(exercise.getName());
            int exerciseCount;
            if ((exerciseCount = mSelectedExercises.get(exercise.getId(), -1)) == -1) {
                mBinding.setSelected(false);
            } else {
                mBinding.setSelected(true);
                mBinding.setCount(exerciseCount);
            }
        }
    }
}

File: app/src/main/java/com/vladislav/workoutassistant/ui/exercises/adapters/ExerciseAdapter.kt
package com.vladislav.workoutassistant.ui.exercises.adapters

import android.util.SparseIntArray
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.vladislav.workoutassistant.R
import com.vladislav.workoutassistant.data.db.entity.Exercise
import com.vladislav.workoutassistant.databinding.ItemExerciseBinding
import com.vladislav.workoutassistant.ui.main.interfaces.OnItemClickCallback
import com.vladislav.workoutassistant.utilities.DiffUtilCallback

class ExerciseAdapter(private val mCallback: OnItemClickCallback, private val mSelectedItems: SparseIntArray) : RecyclerView.Adapter<ExerciseAdapter.ViewHolder>() {

    var lastSelectedItemPosition: Int = 0

    private var mItems: List<Exercise> = listOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ExerciseAdapter.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = DataBindingUtil.inflate<ItemExerciseBinding>(inflater, R.layout.item_exercise, parent, false)
        val holder = ViewHolder(binding)

        holder.itemView.setOnClickListener {
            val item = mItems[holder.adapterPosition]
            mCallback.onClick(item.id, item.name)
            lastSelectedItemPosition = holder.adapterPosition
        }
        return holder
    }

    override fun onBindViewHolder(holder: ExerciseAdapter.ViewHolder, position: Int) {
        val item = mItems[position]
        val amount: Int = mSelectedItems.get(item.id, -1)
        holder.bind(item, amount)
    }

    override fun getItemCount(): Int = mItems.size

    fun updateList(items: List<Exercise>) {
        val diffResult = DiffUtil.calculateDiff(DiffUtilCallback(mItems, items))
        mItems = items
        diffResult.dispatchUpdatesTo(this)
    }



    class ViewHolder(private val mBinding: ItemExerciseBinding) : RecyclerView.ViewHolder(mBinding.root) {

        fun bind(item: Exercise, amount: Int) {
            mBinding.name = item.name
            if (amount >= 0) {
                mBinding.selected = true
                mBinding.count = amount
            } else {
                mBinding.selected = false
            }
        }
    }
}
